compositor_pipeline/src/pipeline/output/whip.rs

```diff
@@ -0,0 +1,226 @@
+    event::Event,
+    pipeline::{AudioCodec, EncoderOutputEvent, PipelineCtx, VideoCodec},
+};
+
+mod establish_peer_connection;
+mod init_peer_connection;
+mod packet_stream;
+mod payloader;
+
+#[derive(Debug)]
+pub struct WhipSender {
+    pub connection_options: WhipSenderOptions,
+    should_close: Arc<AtomicBool>,
+}
+
+#[derive(Debug, Clone)]
+pub struct WhipSenderOptions {
+    pub endpoint_url: String,
+    pub bearer_token: Option<String>,
+    pub video: Option<VideoCodec>,
+    pub audio: Option<AudioCodec>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum WhipError {
+    #[error("Bad status in WHIP response\nStatus: {0}\nBody: {1}")]
+    BadStatus(StatusCode, String),
+
+    #[error("WHIP request failed!\nMethod: {0}\nURL: {1}")]
+    RequestFailed(Method, Url),
+
+    #[error(
+        "Unable to get location endpoint, check correctness of WHIP endpoint and your Bearer token"
+    )]
+    MissingLocationHeader,
+
+    #[error("Invalid endpoint URL: {1}")]
+    InvalidEndpointUrl(#[source] ParseError, String),
+
+    #[error("Missing Host in endpoint URL")]
+    MissingHost,
+
+    #[error("Missing port in endpoint URL")]
+    MissingPort,
+
+    #[error("Failed to create RTC session description: {0}")]
+    RTCSessionDescriptionError(webrtc::Error),
+
+    #[error("Failed to set local description: {0}")]
+    LocalDescriptionError(webrtc::Error),
+
+    #[error("Failed to set remote description: {0}")]
+    RemoteDescriptionError(webrtc::Error),
+
+    #[error("Failed to parse {0} response body: {1}")]
+    BodyParsingError(String, reqwest::Error),
+
+    #[error("Failed to create offer: {0}")]
+    OfferCreationError(webrtc::Error),
+
+    #[error(transparent)]
+    PeerConnectionInitError(#[from] webrtc::Error),
+
+    #[error("Failed to convert ICE candidate to JSON: {0}")]
+    IceCandidateToJsonError(webrtc::Error),
+
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    PayloadingError(#[from] PayloadingError),
+}
+
+impl WhipSender {
+    pub fn new(
+        output_id: &OutputId,
+        options: WhipSenderOptions,
+        packets_receiver: Receiver<EncoderOutputEvent>,
+        request_keyframe_sender: Option<Sender<()>>,
+        pipeline_ctx: &PipelineCtx,
+    ) -> Result<Self, OutputInitError> {
+        let payloader = Payloader::new(options.video, options.audio);
+        let packet_stream = PacketStream::new(packets_receiver, payloader, 1400);
+
+        let should_close = Arc::new(AtomicBool::new(false));
+        let endpoint_url = options.endpoint_url.clone();
+        let bearer_token = options.bearer_token.clone();
+        let output_id = output_id.clone();
+        let should_close2 = should_close.clone();
+        let event_emitter = pipeline_ctx.event_emitter.clone();
+        let request_keyframe_sender = request_keyframe_sender.clone();
+        let tokio_rt = pipeline_ctx.tokio_rt.clone();
+
+        std::thread::Builder::new()
```

Comment:
I don't see any point in adding this thread, you are just starting a thread that waits for another thread to finish.
Just spawn a task here

If I read this correctly you are spawning that thread and register response exits immediately, without waiting for whether the connection is established. I would expect that the request should hang until the connection is established.